### PR TITLE
✨ Modify the userdata secret creation logic

### PR DESCRIPTION
**Pull request by maelk**
**What this PR does / why we need it**:
Currently, we always create a secret with the userdata for BMH. Since CABPK v1alpha3, CABPK creates a secret containing the userData. This PR removes the duplication of the secret where possible.

Several cases:

- no DataSecretName set, but Data is set in the Machine Bootstrap -> creates the secret
- DataSecretName set, but Machine and BMH are in different namespaces -> creates the secret, since the secret is referenced by name only in both cases, assuming that it is in the same namespace as the machine or as the BMH
- DataSecretName set and Machine and BMH in same namespace -> re-use CABPK secret

### CODE REVIEW

File: baremetal/baremetalmachine_manager.go [hunk 1]
```diff
@@ -426,36 +444,43 @@ func (m *MachineManager) Delete(ctx context.Context) error {
...
-			m.setError("Failed to delete BareMetalMachine",
-				capierrors.DeleteMachineError,
-			)
-			return err
+	// Delete created secret, if data was set without DataSecretName or if
+	// BareMetalHost and Machine are in different namespaces.
+	if (m.Machine.Spec.Bootstrap.DataSecretName == nil &&
+		m.Machine.Spec.Bootstrap.Data != nil) ||
+		(m.Machine.Spec.Bootstrap.DataSecretName != nil &&
+			m.Machine.Namespace != host.Namespace) {
```
Comment (kashifest) [inline baremetal/baremetalmachine_manager.go]: This seems a very complex logic, can we simplify it?

File: baremetal/baremetalmachine_manager.go [hunk 2]
```diff
@@ -426,36 +444,43 @@ func (m *MachineManager) Delete(ctx context.Context) error {
...
 				capierrors.DeleteMachineError,
 			)
 			return err
+		} else if err == nil {
+			//unset the finalizers (remove all since we do not expect anything else
+			// to control that object)
+			tmpBootstrapSecret.Finalizers = []string{}
+			err = m.client.Update(ctx, &tmpBootstrapSecret)
+			if err != nil {
+				m.setError("Failed to delete BareMetalMachine",
```
Comment (kashifest) [inline baremetal/baremetalmachine_manager.go]: Should the message contain some info about the actual reason i.e. cannot delete Bootstrap secret, Failed to delete BMM or something like that?

File: baremetal/baremetalmachine_manager.go [hunk 3]
```diff
@@ -426,36 +444,43 @@ func (m *MachineManager) Delete(ctx context.Context) error {
...
+					capierrors.DeleteMachineError,
+				)
+				return err
+			}
+			// Delete the secret with use data
+			err = m.client.Delete(ctx, &tmpBootstrapSecret)
+			if err != nil {
+				m.setError("Failed to delete BareMetalMachine",
```
Comment (kashifest) [inline baremetal/baremetalmachine_manager.go]: Should the message contain some info about the actual reason i.e. cannot delete Bootstrap secret, Failed to delete BMM or something like that?

File: baremetal/baremetalmachine_manager.go (hunk 1 above)
Comment (maelk) [reply baremetal/baremetalmachine_manager.go]: not really. There are two cases, the first one is that objects are not in the same namespace when DataSecretName is set, and the second one is that DataSecretName is not set but Data is set. I don't see how to make it simpler.

**Pull request merged**